File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.py
def solution(m, n, board):
    answer = 0
    moves = [(1,0), (0,1), (1,1)]
    new_board = []
    for i in range(m):
        new_board.append(list(board[i]))
        
    board = new_board
    
    
    def find_upper_block(r, c):
        while r >= 0:
            if board[r][c] != " ":
                return (r,c)
            r -= 1
            
        return (-1,-1)
            
    
    while True:
        total_popped_list = []
        for r in range(m-1):
            for c in range(n-1):
                # 지워진 부분은 pass
                if board[r][c] == " ":
                    continue
                popped_list = [(r,c)]
                isFail = False
                for dr, dc in moves:
                    next_r, next_c = r + dr, c + dc
                    if board[r][c] == board[next_r][next_c]:
                        popped_list.append((next_r,next_c))
                if len(popped_list) == 4:
                    total_popped_list.extend(popped_list)
                    
        if not total_popped_list:
            break
        
        # 지워진 부분 체크
        for r,c in total_popped_list:
            board[r][c] = " "
        
        # 밑으로 쏟아지는 로직
        
        for i in range(n):
            start = m-1
            while start >= 0:
                if board[start][i] == " ":
                    r, c = find_upper_block(start - 1, i)
                    # 블록이 있다면 swap
                    if (r,c) != (-1,-1):
                        board[start][i], board[r][c] = board[r][c], board[start][i]
                    
                start -= 1
                
        
    for i in range(m):
        for j in range(n):
            if board[i][j] == " ":
                answer += 1
                
        
    return answer

File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

pair<int,int> can_change_block(pair<int,int> cur_pos, vector<string> board) {
    
    int cur_r = cur_pos.first;
    int cur_c = cur_pos.second;
    
    for (int next_r = cur_r - 1; next_r >= 0; next_r--) {
        if (board[next_r][cur_c] != '.') {
            return {next_r,cur_c};
        }
    }
    
    return {-1,-1};
}

bool can_pop(pair<int,int> pos, vector<string> board) {
    int dr[4] = {1,1,0};
    int dc[4] = {0,1,1};
    int r = pos.first;
    int c = pos.second;
    char cur_block = board[r][c];
    
    
    
    for (int i = 0; i < 3; i++) {
        int next_r = r + dr[i];
        int next_c = c + dc[i];
        if (cur_block != board[next_r][next_c]) {
            return false;
        }
    }
    
    return true;
}


int solution(int m, int n, vector<string> board) {
    int answer = 0;
    
    while (true) {
        bool result = false;
        vector<pair<int,int>> targets;

        for (int i = 0; i < m - 1; i++) {
            for (int j = 0; j < n - 1; j++) {
                if (board[i][j] == '.') {
                    continue;
                }
                if (can_pop({i,j}, board)) {
                    result = true;
                    targets.push_back({i,j});
                }
            }
        }
        
        for (pair<int,int> target: targets) {
            int target_r = target.first;
            int target_c = target.second;
            board[target_r][target_c] = '.';
            board[target_r+1][target_c] = '.';
            board[target_r+1][target_c+1] = '.';
            board[target_r][target_c+1] = '.';
        }
        
        
        for (int r = m-1; r >= 0; r--) {
            for (int c = 0; c < n; c++) {
                if (board[r][c] == '.') {
                    pair<int,int> swap_pos = can_change_block({r,c}, board);
                    if (swap_pos.first == -1) {
                        continue;
                    }
                    board[r][c] = board[swap_pos.first][swap_pos.second];
                    board[swap_pos.first][swap_pos.second] = '.';
                }
            }
        }
        
        
        
        
        if (!result) {
            break;
        }
    }
    
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            if (board[i][j] == '.') {
                answer += 1;
            }
        }
    }
    
    
    
    return answer;
}